en/tutorials-and-examples/cms/authentication-plugin.rst

```diff
@@ -0,0 +1,288 @@
+        // the infinite redirect loop issue
+        $this->Authentication->addUnauthenticatedActions(['login']);
+    }
+
+    public function login()
+        $this->request->allowMethod(['get', 'post']);
+        $result = $this->Authentication->getResult();
+        // regardless of POST or GET, redirect if user is logged in
+        if ($result->isValid()) {
+            // redirect to /pages/home after login success
+            $redirect = $this->request->getQuery('redirect', [
+                'controller' => 'Pages',
+                'action' => 'display',
+                'home'
+            ]);
+
+            return $this->redirect($redirect);
+        }
+        // display error if user submitted and authentication failed
+        if ($this->request->is('post') && !$result->isValid()) {
+            $this->Flash->error(__('Invalid username or password'));
+        }
+    }
+
+    public function logout()
+    {
+        $result = $this->Authentication->getResult();
+        // regardless of POST or GET, redirect if user is logged in
+        if ($result->isValid()) {
+            $this->Authentication->logout();
+            return $this->redirect(['controller' => 'Users', 'action' => 'login']);
+        }
+    }
+
+Add the template logic for your login action
+
+    // in /templates/Users/login.php
+    <div class="users form">
+        <?= $this->Flash->render('auth') ?>
+        <?= $this->Form->create() ?>
+        <fieldset>
+            <legend><?= __('Please enter your username and password') ?></legend>
+            <?= $this->Form->control('email', ['required' => true]) ?>
+            <?= $this->Form->control('password', ['required' => true]) ?>
+        </fieldset>
+        <?= $this->Form->submit(__('Login')); ?>
+        <?= $this->Form->end() ?>
+    </div>
+
+Now login page will allow us to correctly login into the application.
+Test it by requesting any page of your site and after being redirected
+to the ``/users/login`` page and entering the email and password you
+picked previously when creating your user, you'll be redirected
+successfully after login.
+
+We need to add a couple more details to configure our application,
+we want all ``view`` and ``index`` pages "public" so we'll add this specific
+configuration in AppController.
+
+    // in src/Controller/AppController.php
+    public function beforeFilter(\Cake\Event\EventInterface $event)
+    {
+        parent::beforeFilter($event);
+        // for all controllers in our application, make index and view
+        // actions public, skipping the authentication check
+        $this->Authentication->addUnauthenticatedActions(['index', 'view']);
+    }
+
+.. note::
+
+    If you don't have a user with a hashed password yet, comment the
+    ``loadComponent('Authentication.Authentication')`` line in your AppController
+    and go to ``/users/add`` to create a new user picking email and password.
+    You've just hacked your own app.
```

Comment:
I would not include this line it could confuse people.

Reply:
Let's exchange fun for clarity :)